docs/Overview.md

```diff
@@ -1,96 +1,246 @@
-The difference is that a plan in addition to describing what the operation is, also describes how that operation will be performed.
-In short, a query describes what the operation is and a plan describes how that operation will be carried out.
+type Engine interface {
+    BindValue(id string, Value)
+    Prelude() Scope
+    
+    RegisterSource(id string, CreateSource)
+    RegisterTransformation(id string, CreateTransformation)
+    
+    Query(Representation) (Query, error)
+}
 
-## Relations
+type Scope interface {} // ...
+type Source interface {} // ...
+type Transformation interface {} // ...
+type Value interface {} // ...
+```
 
-Below is a high level description, using the Go language, of the relations between the different components and concepts of the query engine.
+A language is used to describe a query. It consists of words whose letters are taken from an alphabet.
+Letters can be of any type and complexity, from characters, to nodes in an Abstract Syntax Tree (AST). Various languages
+can be used to represent a query.
 
 ```go
-
-type Parser interface {
-    Parse(flux string) Query
+type Language interface {
+    Name() string
 }
+```
 
-// Query defines work to be performed on time series data and a result.
-// A query is represented as a directed acyclic graph (DAG).
+A query defines a unit of work for the engine. It comprises one or more pipelines.
+It contains its representation, its results, and execution statistics.
+
+```go
 type Query interface {
-    Nodes() []QueryNode
-    Edges() []Edge
+    Results() Results
+    Stats() map[string]interface{}
+    Representation() Representation
+}
+
+type Representation interface {
+    Visitor // a representation can be visited for compilation/execution.
+    Lang() Language // a representation makes sense in a certain language.
+    String() string // a visual representation of the query.
+}
+
+// The results for a Query.
+type Results interface {
+    ForEach(func (Result))
 }
 
-// QueryNode is a single step in the DAG of a query
-type QueryNode interface {
-    ID() NodeID
-    // More details about what the node does
+type Result interface {
+    Name() string
+    // ... methods to iterate other the contents of the result
 }
+```
 
-// NodeID uniquely identifies a node.
-type NodeID
+The executor executes a query expressed in some language and fills its results, provided an execution context.
+There can be multiple executors, one for each language supported by the engine.
+The engine is responsible for providing the executor with an
+execution context that embeds everything that is necessary for the execution of the query, included the registered values,
+sources and transformations. 
 
-// Edge establishes a parent child relationship between two nodes.
-type Edge interface {
-    Parent() NodeID
-    Child() NodeID
+```go
+type Executor interface {
+    Execute(Query, ExecutionContext) error // fills query results.
```

Comment:
This interface seems weird to me and it seems more like an implementation rather than a usable interface.

A first problem is that it's unclear how it can use the `Query` interface. There is no way in the public query interface to send results, but only iterate. That would force the interface to use an internal implementation by a typecast to get access at the write components and that will likely result in a bad interface.

I'd like to propose a separate interface that I wrote in some notes when we were discussing this. I had different interface names so I've translated them to their equivalent in this document.

```go
type Executor interface {
    Execute(ExecutionContext, Representation) (Query, error)
    Type() Language
}
```

I did not have the `Type` method in my original proposal, but I like it so I would say to keep it.

Reply:
Absolutely need to add those method to the `Query` interface. Thank you.

I understand your point for `Execute` to accept a `Representation` instead of a `Query`. I took that decision to make executors delegate each other. We should further discuss this point. I see you have a comment on that below.

Reply:
No ok, I was wrong. I agree with changing to `Execute` take a `Representation`.

Reply:
Unfortunately I think I disagree with both. I don't believe an executor should take an `ExecutionContext` as input. Why should it? It is an executor after all. Shouldn't it have all the context that it needs? I had envisioned a much simpler interface like so:
```go
type Executor interface {
    Execute(ir Representation) (Query, error)
}
```